docs(apim): add reporter configuration section (#816)
gravitee-io/gravitee-docs: pages/apim/3.x/installation-guide/configuration/reporters/installation-guide-reporters.adoc

= Reporters
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/apim_installguide_reporters.html
:page-folder: apim/installation-guide/reporters
:page-description: Gravitee.io API Management - Reporters
:page-keywords: Gravitee.io, API Platform, API Management, API Gateway, reporters, events, elasticsearch, json, message_pack, tcp
:page-layout: apim3x

> Reporters are designed to write a variety of events occuring in the gatewy to a variety of outputs and format, giving you the ability to manage this data using a solution of your choice.

Reviewer: ```suggestion
Reporters are designed to write a variety of events occuring in the gateway to a variety of outputs and format, giving you the ability to manage this data using a solution of your choice.
```

Reviewer 2: Good catch @gcusnieux - there was a newer edit in my side but not sure why it was not incorporated, see https://github.com/gravitee-io/gravitee-docs/pull/816#discussion_r907785606 note the additional rewording. We should use this version.

|`node`
|Some system and JVM metrics for the node gravitee is running on.
|===

== Available Reporters

[cols="1,1,1"]
|===
|Type|Bundled|Default 

|Elasticsearch
|✓
|✓

|File
|✓
|×

|TCP
|✓
|×
|===

== Configuring Reporters

=== Elasticsearch Reporter

Configuration details for the elasticsearch reporter are available in the link:{{ '/apim/3.x/apim_installguide_repositories_elasticsearch.html' | relative_url }}[Elasticsearch Repository] documentation.

=== File Reporter

==== Configuration Parameters
[cols="1,1,1"]
|===
| Parameter name
| Description
| Default value

| `enabled`
| Weither to start the file reporter or not
| false

| `fileName`
| File name to write access logs (use '%s-yyyy_mm_dd' pattern to create one file per event type on daily base)
| *#{systemProperties['gravitee.home']}/metrics/%s-yyyy_mm_dd}*

| `output`
| Type of output file (json, message_pack, elasticsearch, csv)
| *json*

| `flushInterval`
| interval between file flush (in ms)
| *1000*

| `retainDays`
| The number of days to retain files before deleting one
| *0 (to retain forever)*

| `<EVENT_TYPE>.exclude`

> | Fields to exclude from the output (available for json` and `message_pack` outputs only)

Reviewer 2: ```suggestion
| Fields to exclude from the output. Available for `json` and `message_pack` outputs only.
```

| *none*

| `<EVENT_TYPE>.rename`
| Fields to rename when writing the output (available for `json`` and `message_pack` outputs only)
| *none*
|===

NOTE: <EVENT_TYPE> refers to the kind of event reported by the gateway and can be either `request`, `log`, `node` or `healthcheck`

==== Example

Here is a configuration example that will exclude all fields from the request JSON file except the `api` and
`application` fields, renaming the `application` field to `app`, and will not report `log`, `node` or `healthcheck` events

[source,yaml]
----
reporters:
  file:
    enabled: true 
    fileName: ${gravitee.home}/metrics/%s-yyyy_mm_dd
    output: json 
    request: 
      exclude: 
        - "*"
      include: 
        - api
        - application
      rename: 
        application: app
    log:
      exclude: *
    node:
      exclude: *
    healthcheck:
      exclude: *
----

NOTE: the fileName property **must** end with the `%s-yyyy_mm_dd` pattern, otherwise your file will not be written.

=== TCP Reporter

==== Configuration Parameters
[cols="1,1,1"]
|===
| Parameter name
| Description
| Default value

| `enabled`

> | Weither to start the TCP reporter or not

Reviewer 2: ```suggestion
| This setting determines whether the TCP reporter should be started or not. The default value is `false`.
```

| false

| `output`
| Format of the data written to the TCP socket (json, message_pack, elasticsearch, csv)
| *json*

| `host`

> | The TCP host where the event should be published (can be a valid host name or IP address)

Reviewer 2: ```suggestion
| The TCP host where the event should be published. This can be a valid host name or an IP address.
```

| *localhost*

| `port`

> | The TCP port used to connect to the host

Reviewer 2: ```suggestion
| The TCP port used to connect to the host.
```

| *8123*

| `connectTimeout`

> | Maximum time allowed to establish the TCP connection in milliseconds

Reviewer 2: ```suggestion
| Maximum time allowed to establish the TCP connection in milliseconds.
```

| *10000*

| `reconnectAttempts`

> | How many time the socket should try to establish a connection in case of failure

Reviewer 2: ```suggestion
| This setting determines how many times the socket should try to establish a connection in case of failure.
```

| *10*

| `reconnectInterval`

> | Time in millisceconds between each socket connection attempt

Reviewer 2: ```suggestion
| Time (in milliseconds) between socket connection attempts.
```

| *500*

| `retryTimeout`

> | If the max reconnect attempts have been reached, how many time in milliseconds the reporter should wait before trying to connect again

Reviewer 2: ```suggestion
| If the max reconnect attempts have been reached, this setting determines how many times (in milliseconds) the reporter should wait before trying to connect again.
```

| *5000*

|===

==== Example

> We can define the same configuration as above, but write the events to a TCP socket instead of a file using

Reviewer 2: ```suggestion
The following example uses the same configuration as the previous example above, however it writes the events to a TCP socket instead of a file.
```

> the following configuration

Reviewer 2: ```suggestion

```

Reviewer 2: Sorry I meant to say we should remove this line as the main sentence has been reworded.
